lib/silence_generator.ex

```diff
@@ -49,33 +49,55 @@ defmodule Membrane.SilenceGenerator do
   @impl true
   def handle_demand(:output, size, :bytes, _ctx, %{caps: caps} = state) do
     time = RawAudio.bytes_to_time(size, caps)
-    do_handle_demand(time, state)
+    do_handle_demand(time, time, state)
   end
 
   def handle_demand(:output, buffers, :buffers, _ctx, state) do
     %{caps: caps, frames_per_buffer: frames_per_buffer} = state
 
-    time = buffers * RawAudio.frames_to_time(frames_per_buffer, caps)
-    do_handle_demand(time, state)
+    time = RawAudio.frames_to_time(frames_per_buffer, caps)
+    do_handle_demand(time * buffers, time, state)
   end
 
   defp do_handle_demand(
-         time,
+         total_time,
+         chunk_time,
          %{caps: caps, duration: :infinity, passed_time: passed_time} = state
        ) do
-    buffer = %Buffer{payload: RawAudio.silence(caps, time), pts: passed_time}
-    {{:ok, buffer: {:output, buffer}}, %{state | passed_time: passed_time + time}}
+    buffers = generate_buffers(passed_time, chunk_time, total_time, caps)
+    state = %{state | passed_time: passed_time + total_time}
+
+    {{:ok, buffer: {:output, buffers}}, state}
+  end
+
+  defp do_handle_demand(
+         total_time,
+         chunk_time,
+         %{caps: caps, duration: duration, passed_time: passed_time} = state
+       ) do
+    buffers =
+      generate_buffers(
+        passed_time,
+        chunk_time,
+        min(total_time, duration - passed_time),
+        caps
+      )
+
+    state = %{state | passed_time: min(passed_time + total_time, duration)}
+
+    if state.passed_time == duration,
+      do: {{:ok, buffer: {:output, buffers}, end_of_stream: :output}, state},
+      else: {{:ok, buffer: {:output, buffers}}, state}
   end
 
-  defp do_handle_demand(time, state) do
-    %{caps: caps, duration: duration, passed_time: passed_time} = state
+  defp generate_buffers(start_time, chunk_time, total_time, caps) do
+    Enum.map(0..floor(total_time / chunk_time), fn x ->
```

Comment:
I think this would actually look better as a recursion :D But in case you prefer to leave it as is:
```suggestion
    Enum.map(0..floor(total_time / chunk_time)//1, fn i ->
```